File: src/main/kotlin/com/marcel/baumgardt/model/mapper/ShowingDatesResponseMapper.java
package com.marcel.baumgardt.model.mapper;

import com.marcel.baumgardt.model.db.Showing;
import com.marcel.baumgardt.model.dto.ShowingDate;
import com.marcel.baumgardt.model.dto.ShowingDatesResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.time.OffsetTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.groupingBy;

@Component
@RequiredArgsConstructor
public class ShowingDatesResponseMapper {
    public ShowingDatesResponse mapToShowingDatesResponse(Long cinemaId, Long movieId, List<Showing> showings) {
        List<ShowingDate> showingDates = showings.stream()
                .collect(groupingBy(Showing::getDay))
                .entrySet()
                .stream()
                .map(this::mapToShowingDates)
                .collect(Collectors.toList());

        return ShowingDatesResponse.builder()
                .cinemaId(cinemaId)
                .movieId(movieId)
                .dates(showingDates)
                .build();
    }

    private ShowingDate mapToShowingDates(Map.Entry<Integer, List<Showing>> groupedShowingsEntry) {
        return ShowingDate.builder()
                .day(groupedShowingsEntry.getKey())
                .timeList(mapToTimeList(groupedShowingsEntry.getValue()))
                .build();
    }

    private List<OffsetTime> mapToTimeList(List<Showing> showings) {
        return showings.stream()
                .map(Showing::getTime)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/marcel/baumgardt/model/mapper/ShowingDatesResponseMapper.kt
package com.marcel.baumgardt.model.mapper

import com.marcel.baumgardt.model.db.Showing
import com.marcel.baumgardt.model.dto.common.ShowingDate
import com.marcel.baumgardt.model.dto.response.ShowingDatesResponse
import org.springframework.stereotype.Component
import java.time.OffsetTime
import java.util.stream.Collectors.groupingBy

@Component
class ShowingDatesResponseMapper {

    fun mapToShowingDatesResponse(cinemaId: Long, movieId: Long, showings: List<Showing>): ShowingDatesResponse {
        val showingDates = showings.stream().collect(groupingBy { s -> s.day })
            .entries
            .stream()
            .map { e -> mapToShowingDates(e) }
            .toList()

        val cinemaName = showings.first().cinema.name
        val movieTitle = showings.first().movie.title

        return ShowingDatesResponse(
            cinemaId,
            cinemaName,
            movieId,
            movieTitle,
            showingDates
        )
    }

    private fun mapToShowingDates(groupedShowingsEntry: Map.Entry<Int, List<Showing>>): ShowingDate {
        return ShowingDate(
            groupedShowingsEntry.key,
            mapToTimeList(groupedShowingsEntry.value)
        )
    }

    private fun mapToTimeList(showings: List<Showing>): List<OffsetTime> {
        return showings
            .stream()
            .map { s -> s.time }
            .toList()
    }
}